### PR TITLE
mds: require config list to be sorted

### PR DESCRIPTION
**Pull request by batrick**
Fun little programming project for the obsessive mind. I find it easier to have the configs grouped by component (part of the config name usually) so keeping the alphabetized is helpful. It also helps to avoid conflicts if everyone would otherwise be appending to the end of the array.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
jenkins test api

**Comment by vshankar** [comment]
jenkins test api

**Comment by batrick** [comment]
```
ninja: build stopped: interrupted by user.
Build step 'Execute shell' marked build as failure
```

:open_mouth:

**Comment by batrick** [comment]
jenkins test api

File: src/common/make_array.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+#pragma once
+
+#include <type_traits>
+#include <array>
+
+template<typename CT, typename... Args>
+constexpr auto make_array(Args&&... args)
+{
+  return std::array<CT, sizeof...(Args)>{std::forward<CT>(args)...};
```
Comment (dparmar18) [inline src/common/make_array.h:23]: Shouldn't we be forwarding args with type `Args`?
```suggestion
  return std::array<CT, sizeof...(Args)>{std::forward<Args>(args)...};
```

**Comment by vshankar** [comment]
> Fun little programming project for the obsessive mind

Keeps the neurons firing :)

Anyway, I'm fine with having this constraint in place. Follow up question unrelated to this change: Is it possible to auto-generate this list of keys from mds.yaml..in by any chance?

File: src/common/make_array.h (hunk 1 above)
Comment (batrick) [reply src/common/make_array.h]: I've since found `std::to_array` so code no longer exists.

But to answer your question: I think so.

**Comment by batrick** [comment]
> > Fun little programming project for the obsessive mind
> 
> Keeps the neurons firing :)
> 
> Anyway, I'm fine with having this constraint in place. Follow up question unrelated to this change: Is it possible to auto-generate this list of keys from mds.yaml..in by any chance?

I think the idea of this interface was to prevent unnecessary "config change" deliveries to the daemon. I personally don't see the issue unless you are acquiring expensive locks but that can be avoided usually. So to answer your question: I think that adding all the keys from `mds.yaml.in` would be abusing the config API.

**Comment by vshankar** [comment]
> > > Fun little programming project for the obsessive mind
> > 
> > 
> > Keeps the neurons firing :)
> > Anyway, I'm fine with having this constraint in place. Follow up question unrelated to this change: Is it possible to auto-generate this list of keys from mds.yaml..in by any chance?
> 
> I think the idea of this interface was to prevent unnecessary "config change" deliveries to the daemon. I personally don't see the issue unless you are acquiring expensive locks but that can be avoided usually. So to answer your question: I think that adding all the keys from `mds.yaml.in` would be abusing the config API.

Oh, I meant something in compile/build time that creates a cc source with keys from mds.yaml.in.

**Review by dparmar18 — APPROVED** [review]
LGTM

**Comment by batrick** [comment]
> > > > Fun little programming project for the obsessive mind
> > > 
> > > 
> > > Keeps the neurons firing :)
> > > Anyway, I'm fine with having this constraint in place. Follow up question unrelated to this change: Is it possible to auto-generate this list of keys from mds.yaml..in by any chance?
> > 
> > 
> > I think the idea of this interface was to prevent unnecessary "config change" deliveries to the daemon. I personally don't see the issue unless you are acquiring expensive locks but that can be avoided usually. So to answer your question: I think that adding all the keys from `mds.yaml.in` would be abusing the config API.
> 
> Oh, I meant something in compile/build time that creates a cc source with keys from mds.yaml.in.

I'm not sure I follow what you're trying to accomplish by doing so?

**Comment by vshankar** [comment]
> > > > > Fun little programming project for the obsessive mind
> > > > 
> > > > 
> > > > Keeps the neurons firing :)
> > > > Anyway, I'm fine with having this constraint in place. Follow up question unrelated to this change: Is it possible to auto-generate this list of keys from mds.yaml..in by any chance?
> > > 
> > > 
> > > I think the idea of this interface was to prevent unnecessary "config change" deliveries to the daemon. I personally don't see the issue unless you are acquiring expensive locks but that can be avoided usually. So to answer your question: I think that adding all the keys from `mds.yaml.in` would be abusing the config API.
> > 
> > 
> > Oh, I meant something in compile/build time that creates a cc source with keys from mds.yaml.in.
> 
> I'm not sure I follow what you're trying to accomplish by doing so?

Trying to have a single place where we specify config options and then automagically have the list of keys in MDS source.

Never mind - that's my obsessive mind working late evenings.

**Comment by batrick** [comment]
> > > > > > Fun little programming project for the obsessive mind
> > > > > 
> > > > > 
> > > > > Keeps the neurons firing :)
> > > > > Anyway, I'm fine with having this constraint in place. Follow up question unrelated to this change: Is it possible to auto-generate this list of keys from mds.yaml..in by any chance?
> > > > 
> > > > 
> > > > I think the idea of this interface was to prevent unnecessary "config change" deliveries to the daemon. I personally don't see the issue unless you are acquiring expensive locks but that can be avoided usually. So to answer your question: I think that adding all the keys from `mds.yaml.in` would be abusing the config API.
> > > 
> > > 
> > > Oh, I meant something in compile/build time that creates a cc source with keys from mds.yaml.in.
> > 
> > 
> > I'm not sure I follow what you're trying to accomplish by doing so?
> 
> Trying to have a single place where we specify config options and then automagically have the list of keys in MDS source.
> 
> Never mind - that's my obsessive mind working late evenings.

ahhh, that would be interesting. Actually, you could just pull the list of configs with the `RUNTIME` flag.

**Comment by vshankar** [comment]
> > > > > > > Fun little programming project for the obsessive mind
> > > > > > 
> > > > > > 
> > > > > > Keeps the neurons firing :)
> > > > > > Anyway, I'm fine with having this constraint in place. Follow up question unrelated to this change: Is it possible to auto-generate this list of keys from mds.yaml..in by any chance?
> > > > > 
> > > > > 
> > > > > I think the idea of this interface was to prevent unnecessary "config change" deliveries to the daemon. I personally don't see the issue unless you are acquiring expensive locks but that can be avoided usually. So to answer your question: I think that adding all the keys from `mds.yaml.in` would be abusing the config API.
> > > > 
> > > > 
> > > > Oh, I meant something in compile/build time that creates a cc source with keys from mds.yaml.in.
> > > 
> > > 
> > > I'm not sure I follow what you're trying to accomplish by doing so?
> > 
> > 
> > Trying to have a single place where we specify config options and then automagically have the list of keys in MDS source.
> > Never mind - that's my obsessive mind working late evenings.
> 
> ahhh, that would be interesting. Actually, you could just pull the list of configs with the `RUNTIME` flag.
> 
> Edit: and you'd want to indicate which RUNTIME configs are watched for which daemon.

Agree. You don't have to do that in this PR though. I'll create a ticket for our enthusiasts :)

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/67563](https://tracker.ceph.com/issues/67563).

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
https://tracker.ceph.com/issues/67563#note-1

**Pull request merged**